tools/stronghold/src/api/compatibility.py

```diff
@@ -0,0 +1,148 @@
+"""Understands compatibility of APIs."""
+
+from __future__ import annotations
+
+from collections.abc import Iterable, Mapping, Sequence
+import dataclasses
+import itertools
+import pathlib
+
+import api
+import api.ast
+
+
+def check(before: pathlib.Path, after: pathlib.Path) -> Sequence[Violation]:
+    """Identifies API compatibility issues between two files."""
+    before_api = api.ast.extract(before)
+    after_api = api.ast.extract(after)
+
+    violations = []
+    for name, before_def in before_api.items():
+        if any(token.startswith('_') for token in name.split('.')):
+            continue
+
+        after_def = after_api.get(name)
+        if after_def is None:
+            violations.append(Violation(name, 'function deleted'))
+            continue
+
+        # Let's refine some terminology. Parameters come in three flavors:
+        #  * positional only
+        #  * keyword only
+        #  * flexible: may be provided positionally or via keyword
+        #
+        # Required parameter: a parameter that must be provided as an
+        # argument by callers. In other words, the function does not
+        # have a default value.
+        #
+        # Variadic parameters: additional arguments that may only be
+        # provided positionally, traditionally specified as *args in a
+        # function definition.
+        #
+        # Variadic keywords: additional arguments that may only be
+        # provided by name, traditionally specified as **kwargs in a
+        # function definition.
+
+        violations += _check_by_name(name, before_def, after_def)
+        violations += _check_by_position(name, before_def, after_def)
+        violations += _check_variadic_parameters(name, before_def, after_def)
+
+    return violations
+
+
+@dataclasses.dataclass
```

Comment:
nit: eventually separate dataclasses from code?

Reply:
FWIW, i like to keep my types coupled to the code that operates on them principally. So I like the idea of a Violation being part of the compatibility module, since a Violation is inherently about compatibility.

I pulled the Parameter types into the API module since they are fundamental to all the code and there may be other extractors than just the ast one.